fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTask.java

```diff
@@ -0,0 +1,101 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.task;
+
+import com.google.common.base.Preconditions;
+import com.starrocks.sql.optimizer.ExpressionContext;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.Rule;
+
+import java.util.List;
+
+public class RewriteTask extends OptimizerTask {
+    private final OptExpression planTree;
+    private final boolean onlyOnce;
+    private final List<Rule> rules;
+    private long change = 0;
+
+    public RewriteTask(TaskContext context, OptExpression root, List<Rule> rules, boolean onlyOnce) {
+        super(context);
+        this.planTree = root;
+        this.rules = rules;
+        this.onlyOnce = onlyOnce;
+    }
+
+    public OptExpression getResult() {
+        return planTree.getInputs().get(0);
+    }
+
+    @Override
+    public void execute() {
+        // first node must be RewriteAnchorNode
+        rewrite(planTree, 0, planTree.getInputs().get(0));
+
+        if (change > 0 && !onlyOnce) {
+            pushTask(new RewriteTask(context, planTree, rules, onlyOnce));
+        }
+    }
+
+    private void rewrite(OptExpression parent, int childIndex, OptExpression root) {
+        for (Rule rule : rules) {
+            if (!match(rule.getPattern(), root) || !rule.check(root, context.getOptimizerContext())) {
```

Comment:
if dosen't meet pattern, should we skip this rule directly and there is no need to use rule.check() ?